File: LeetCode-September-Challenge/DesignCircularQueue.java
import java.util.Arrays;

class MyCircularQueue {
    private int rear = 0, front = 0, size = 0;
    private int[] arr;
    public MyCircularQueue(int k) {
        arr = new int[k];
        Arrays.fill(arr, -1);
    }
    
    public boolean enQueue(int value) {
        if(isFull()) return false;
        if(isEmpty()){
            rear = front = 0;
            arr[rear] = value;
            size++;
            return true;
        }
        rear++;
        size++;
        rear %= arr.length;
        arr[rear] = value;
        return true;
    }
    
    public boolean deQueue() {
        if(isEmpty()) return false;
        arr[front] = -1;
        front++;
        size--;
        front %= arr.length;
        return true;
    }
    
    public int Front() {
        return arr[front];
    }
    
    public int Rear() {
        return arr[rear];
    }
    
    public boolean isEmpty() {
        return size == 0;
    }
    
    public boolean isFull() {
        return size == arr.length;
    }
}
public class DesignCircularQueue {
    public static void main(String[] args) {
        MyCircularQueue q = new MyCircularQueue(3);
        System.out.println(q.enQueue(1));
        System.out.println(q.enQueue(2));
        System.out.println(q.enQueue(3));
        System.out.println(q.enQueue(4));
        System.out.println(q.Rear());
        System.out.println(q.isFull());
        System.out.println(q.deQueue());
        System.out.println(q.enQueue(4));
        System.out.println(q.Rear());
    }
}

File: Year-2022/LeetCode-September-Challenge/DesignCircularQueue.cpp
#include<bits/stdc++.h>
using namespace std;
class MyCircularQueue {
private:
    int rear = 0;
    int front = 0;
    int size = 0;
    int capacity = 0;
    vector<int> arr;
public:
    MyCircularQueue(int k) {
        capacity = k;
        arr.resize(k);
    }
    
    bool enQueue(int value) {
        if(isFull()) return false;
        if(isEmpty()){
            front = rear = 0;
            arr[rear] = value;
            size++;
            return true;
        }
        rear++;
        size++;
        rear %= capacity;
        arr[rear] = value;
        return true;
    }
    
    bool deQueue() {
        if(isEmpty()) return false;
        arr[front] = -1;
        size--;
        front++;
        front %= capacity;
        return true;
    }
    int Front() {
        return isEmpty() ? -1 : arr[front];
    }
    
    int Rear() {
        return isEmpty() ? -1 : arr[rear];
    }
    
    bool isEmpty() {
        return size == 0;
    }
    
    bool isFull() {
        return size == capacity;
    }
};
int main()
{
    MyCircularQueue q(3);
    cout << q.enQueue(1) << endl;
    cout << q.enQueue(2) << endl;
    cout << q.enQueue(3) << endl;
    cout << q.enQueue(4) << endl;
    cout << q.Rear() << endl;
    cout << q.isFull() << endl;
    cout << q.deQueue() << endl;
    cout << q.enQueue(4) << endl;
    cout << q.Rear() << endl;
    return 0;
}